applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_adjoint_static_solver.py

```diff
@@ -0,0 +1,127 @@
+
+    def PrepareModelPart(self):
+        if(self.main_model_part.ProcessInfo[KratosMultiphysics.DOMAIN_SIZE]!= 3):
+            raise Exception("there are currently only 3D adjoint elements available")
+        super(StructuralMechanicsAdjointStaticSolver, self).PrepareModelPart()
+        # TODO Why does replacement need to happen after reading materials?
+        StructuralMechanicsApplication.ReplaceElementsAndConditionsForAdjointProblemProcess(self.main_model_part).Execute()
+        self.print_on_rank_zero("::[AdjointMechanicalSolver]:: ", "ModelPart prepared for Solver.")
+
+    def AddDofs(self):
+        KratosMultiphysics.VariableUtils().AddDof(StructuralMechanicsApplication.ADJOINT_DISPLACEMENT_X, self.main_model_part)
+        KratosMultiphysics.VariableUtils().AddDof(StructuralMechanicsApplication.ADJOINT_DISPLACEMENT_Y, self.main_model_part)
+        KratosMultiphysics.VariableUtils().AddDof(StructuralMechanicsApplication.ADJOINT_DISPLACEMENT_Z, self.main_model_part)
+        if self.settings["rotation_dofs"].GetBool():
+            KratosMultiphysics.VariableUtils().AddDof(StructuralMechanicsApplication.ADJOINT_ROTATION_X, self.main_model_part)
+            KratosMultiphysics.VariableUtils().AddDof(StructuralMechanicsApplication.ADJOINT_ROTATION_Y, self.main_model_part)
+            KratosMultiphysics.VariableUtils().AddDof(StructuralMechanicsApplication.ADJOINT_ROTATION_Z, self.main_model_part)
+        self.print_on_rank_zero("::[AdjointMechanicalSolver]:: ", "DOF's ADDED.")
+
+    def Initialize(self):
+        """Perform initialization after adding nodal variables and dofs to the main model part. """
+        if self.response_function_settings["response_type"].GetString() == "adjoint_local_stress":
+            self.response_function = StructuralMechanicsApplication.AdjointLocalStressResponseFunction(self.main_model_part, self.response_function_settings)
+        elif self.response_function_settings["response_type"].GetString() == "adjoint_nodal_displacement":
+            self.response_function = StructuralMechanicsApplication.AdjointNodalDisplacementResponseFunction(self.main_model_part, self.response_function_settings)
+        elif self.response_function_settings["response_type"].GetString() == "adjoint_strain_energy":
+            self.response_function = StructuralMechanicsApplication.AdjointStrainEnergyResponseFunction(self.main_model_part, self.response_function_settings)
+        else:
+            raise Exception("invalid response_type: " + self.response_function_settings["response_type"].GetString())
+
+        super(StructuralMechanicsAdjointStaticSolver, self).Initialize()
+
+        self.print_on_rank_zero("::[AdjointMechanicalSolver]:: ", "Finished initialization.")
+
+    def Solve(self):
+        if self.response_function_settings["response_type"].GetString() == "adjoint_strain_energy":
+            self._SolveSolutionStepSpecialStrainEnergy()
+        else:
```

Comment:
why not call the baseclass instead of copy pasting code?

Edit I think you should not implement `Solve` at all
=> it should be `SolveSolutionStep

We should discuss it`

Reply:
Refactored a little